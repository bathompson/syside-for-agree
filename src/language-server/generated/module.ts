/********************************************************************************
 * Copyright (c) 2022-2023 Sensmetry UAB and others
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License, v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is
 * available at https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/

/******************************************************************************
 * This file was generated by langium-cli 0.5.0.
 * DO NOT EDIT MANUALLY!
 ******************************************************************************/

import { LangiumGeneratedServices, LangiumGeneratedSharedServices, LangiumSharedServices, LangiumServices, LanguageMetaData, Module, IParserConfig } from 'langium';
import { SysMlAstReflection } from './ast';
import { KerMLGrammar, SysMLGrammar } from './grammar';

export const KerMLLanguageMetaData: LanguageMetaData = {
    languageId: 'kerml',
    fileExtensions: ['.kerml'],
    caseInsensitive: false
};

export const SysMLLanguageMetaData: LanguageMetaData = {
    languageId: 'sysml',
    fileExtensions: ['.sysml'],
    caseInsensitive: false
};

export const parserConfig: IParserConfig = {
    recoveryEnabled: true,
    nodeLocationTracking: 'none',
};

export const SysMlGeneratedSharedModule: Module<LangiumSharedServices, LangiumGeneratedSharedServices> = {
    AstReflection: () => new SysMlAstReflection()
};

export const KerMLGeneratedModule: Module<LangiumServices, LangiumGeneratedServices> = {
    Grammar: () => KerMLGrammar(),
    LanguageMetaData: () => KerMLLanguageMetaData,
    parser: {
        ParserConfig: () => parserConfig
    }
};

export const SysMLGeneratedModule: Module<LangiumServices, LangiumGeneratedServices> = {
    Grammar: () => SysMLGrammar(),
    LanguageMetaData: () => SysMLLanguageMetaData,
    parser: {
        ParserConfig: () => parserConfig
    }
};
