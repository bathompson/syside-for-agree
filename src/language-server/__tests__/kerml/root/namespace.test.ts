/********************************************************************************
 * Copyright (c) 2022-2023 Sensmetry UAB and others
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License, v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is
 * available at https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/

import { qualifiedTypeReference } from "../../../../testing";
import { Relationship, Class, DataType, Package, Feature } from "../../../generated/ast";
import { Element } from "../../../generated/ast";

test("namespaces are parseable", async () => {
    return expect(`namespace <'1.1'> N1; // This is an empty namespace.
    namespace <'1.2'> N2 {
    doc /* This is an example of a namespace body. */
    class C;
    datatype D;
    feature f : C;
    namespace N3; // This is a nested namespace.
    }`).toParseKerML({
        elements: [
            { name: "N1", shortName: "'1.1'" },
            {
                name: "N2",
                shortName: "'1.2'",
                docs: [{ body: "/* This is an example of a namespace body. */" }],
                elements: [{ name: "C" }, { name: "D" }, { name: "N3" }],
                features: [{ name: "f", typedBy: [qualifiedTypeReference("N2::C")] }],
            },
        ],
    });
});

test("namespace elements can have visibility", async () => {
    return expect(`namespace N {
        public class C;
        private datatype D;
        protected feature f : C;
    }`).toParseKerML({
        elements: [
            {
                name: "N",
                elements: [
                    { name: "C", visibility: "public" },
                    { name: "D", visibility: "private" },
                ],
                features: [
                    {
                        name: "f",
                        visibility: "protected",
                        typedBy: [qualifiedTypeReference("N::C")],
                    },
                ],
            },
        ],
    });
});

test("aliases are parseable", async () => {
    return expect(`namespace N1 {
        class A;
        class B;
        alias <C> CCC for B {
            doc /* alias doc */
        }
        private alias D for B;
    }`).toParseKerML({
        elements: [
            {
                name: "N1",
                elements: [{ name: "A" }, { name: "B" }],
                aliases: [
                    {
                        name: "CCC",
                        shortName: "C",
                        for: qualifiedTypeReference("N1::B"),
                        docs: [{ body: "/* alias doc */" }],
                    },
                    { name: "D", for: qualifiedTypeReference("N1::B"), visibility: "private" },
                ],
            },
        ],
    });
});

test("namespaces can have comments", async () => {
    return expect(`namespace N9 {
    class A;
    comment Comment1 about A
        /* comment about A */
    comment Comment2
        /* comment about N9 */
    /* also comment about N9 */
    doc N9_Doc
        /* doc about N9 */
    }`).toParseKerML({
        elements: [
            {
                name: "N9",
                elements: [{ name: "A" }],
                comments: [
                    {
                        name: "Comment1",
                        about: [qualifiedTypeReference("N9::A")],
                        body: "/* comment about A */",
                    },
                    { name: "Comment2", body: "/* comment about N9 */" },
                    { body: "/* also comment about N9 */" },
                ],
                docs: [{ name: "N9_Doc", body: "/* doc about N9 */" }],
            },
        ],
    });
});

test("all top-level elements are in the root namespace", async () => {
    return expect(`doc /* root doc */
    element A {
        relationship B to C;
    }
    class C;
    datatype D;
    feature f: C;
    package P;
    `).toParseKerML({
        docs: [{ $type: "Documentation", body: "/* root doc */" }],
        elements: [
            {
                $type: Element,
                name: "A",
                relationships: [
                    {
                        $type: Relationship,
                        name: "B",
                        target: [qualifiedTypeReference("C")],
                    },
                ],
            },
            { name: "C", $type: Class },
            { name: "D", $type: DataType },
            { name: "P", $type: Package },
        ],
        features: [{ name: "f", $type: Feature }],
    });
});
