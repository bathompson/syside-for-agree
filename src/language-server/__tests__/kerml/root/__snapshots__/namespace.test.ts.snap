// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aliases are parseable: toParseKerML 1`] = `
{
  "diagnostics": [],
  "lexerErrors": [],
  "parserErrors": [],
  "value": {
    "$type": "Namespace",
    "aliases": [],
    "annotatingMembers": [],
    "filters": [],
    "imports": [],
    "members": [],
    "namespaceMembers": [
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Namespace",
          "aliases": [
            {
              "$type": "Membership",
              "annotations": [
                {
                  "$type": "Annotation",
                  "annotations": [],
                  "chains": [],
                  "element": {
                    "$type": "Documentation",
                    "about": [],
                    "body": "/* alias doc */",
                    "prefixes": [],
                  },
                  "elements": [],
                  "prefixes": [],
                },
              ],
              "chains": [],
              "declaredName": "CCC",
              "declaredShortName": "C",
              "elements": [],
              "prefixes": [],
              "reference": {
                "$type": "ElementReference",
                "parts": [
                  {
                    "error": undefined,
                    "ref": {
                      "$type": "Class",
                      "aliases": [],
                      "annotatingMembers": [],
                      "declaredName": "B",
                      "filters": [],
                      "imports": [],
                      "isSufficient": false,
                      "members": [],
                      "namespaceMembers": [],
                      "prefixes": [],
                      "relationshipMembers": [],
                      "typeRelationships": [],
                    },
                  },
                ],
              },
            },
            {
              "$type": "Membership",
              "annotations": [],
              "chains": [],
              "declaredName": "D",
              "elements": [],
              "prefixes": [],
              "reference": {
                "$type": "ElementReference",
                "parts": [
                  {
                    "error": undefined,
                    "ref": {
                      "$type": "Class",
                      "aliases": [],
                      "annotatingMembers": [],
                      "declaredName": "B",
                      "filters": [],
                      "imports": [],
                      "isSufficient": false,
                      "members": [],
                      "namespaceMembers": [],
                      "prefixes": [],
                      "relationshipMembers": [],
                      "typeRelationships": [],
                    },
                  },
                ],
              },
              "visibility": "private",
            },
          ],
          "annotatingMembers": [],
          "declaredName": "N1",
          "filters": [],
          "imports": [],
          "members": [],
          "namespaceMembers": [
            {
              "$type": "OwningMembership",
              "annotations": [],
              "chains": [],
              "element": {
                "$type": "Class",
                "aliases": [],
                "annotatingMembers": [],
                "declaredName": "A",
                "filters": [],
                "imports": [],
                "isSufficient": false,
                "members": [],
                "namespaceMembers": [],
                "prefixes": [],
                "relationshipMembers": [],
                "typeRelationships": [],
              },
              "elements": [],
              "prefixes": [],
            },
            {
              "$type": "OwningMembership",
              "annotations": [],
              "chains": [],
              "element": {
                "$type": "Class",
                "aliases": [],
                "annotatingMembers": [],
                "declaredName": "B",
                "filters": [],
                "imports": [],
                "isSufficient": false,
                "members": [],
                "namespaceMembers": [],
                "prefixes": [],
                "relationshipMembers": [],
                "typeRelationships": [],
              },
              "elements": [],
              "prefixes": [],
            },
          ],
          "prefixes": [],
          "relationshipMembers": [],
        },
        "elements": [],
        "prefixes": [],
      },
    ],
    "prefixes": [],
    "relationshipMembers": [],
  },
}
`;

exports[`all top-level elements are in the root namespace: toParseKerML 1`] = `
{
  "diagnostics": [],
  "lexerErrors": [],
  "parserErrors": [],
  "value": {
    "$type": "Namespace",
    "aliases": [],
    "annotatingMembers": [
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Documentation",
          "about": [],
          "body": "/* root doc */",
          "prefixes": [],
        },
        "elements": [],
        "prefixes": [],
      },
    ],
    "filters": [],
    "imports": [],
    "members": [
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Feature",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "f",
          "filters": [],
          "imports": [],
          "isNonunique": false,
          "isOrdered": false,
          "isSufficient": false,
          "members": [],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [
            {
              "$type": "FeatureTyping",
              "annotations": [],
              "chains": [],
              "elements": [],
              "prefixes": [],
              "reference": {
                "$type": "TypeReference",
                "parts": [
                  {
                    "error": undefined,
                    "ref": {
                      "$type": "Class",
                      "aliases": [],
                      "annotatingMembers": [],
                      "declaredName": "C",
                      "filters": [],
                      "imports": [],
                      "isSufficient": false,
                      "members": [],
                      "namespaceMembers": [],
                      "prefixes": [],
                      "relationshipMembers": [],
                      "typeRelationships": [],
                    },
                  },
                ],
              },
            },
          ],
        },
        "elements": [],
        "prefixes": [],
      },
    ],
    "namespaceMembers": [
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Class",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "A",
          "filters": [],
          "imports": [],
          "isSufficient": false,
          "members": [],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [],
        },
        "elements": [],
        "prefixes": [],
      },
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Class",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "C",
          "filters": [],
          "imports": [],
          "isSufficient": false,
          "members": [],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [],
        },
        "elements": [],
        "prefixes": [],
      },
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "DataType",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "D",
          "filters": [],
          "imports": [],
          "isSufficient": false,
          "members": [],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [],
        },
        "elements": [],
        "prefixes": [],
      },
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Package",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "P",
          "filters": [],
          "imports": [],
          "members": [],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
        },
        "elements": [],
        "prefixes": [],
      },
    ],
    "prefixes": [],
    "relationshipMembers": [],
  },
}
`;

exports[`namespace elements can have visibility: toParseKerML 1`] = `
{
  "diagnostics": [],
  "lexerErrors": [],
  "parserErrors": [],
  "value": {
    "$type": "Namespace",
    "aliases": [],
    "annotatingMembers": [],
    "filters": [],
    "imports": [],
    "members": [],
    "namespaceMembers": [
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Namespace",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "N",
          "filters": [],
          "imports": [],
          "members": [
            {
              "$type": "OwningMembership",
              "annotations": [],
              "chains": [],
              "element": {
                "$type": "Feature",
                "aliases": [],
                "annotatingMembers": [],
                "declaredName": "f",
                "filters": [],
                "imports": [],
                "isNonunique": false,
                "isOrdered": false,
                "isSufficient": false,
                "members": [],
                "namespaceMembers": [],
                "prefixes": [],
                "relationshipMembers": [],
                "typeRelationships": [
                  {
                    "$type": "FeatureTyping",
                    "annotations": [],
                    "chains": [],
                    "elements": [],
                    "prefixes": [],
                    "reference": {
                      "$type": "TypeReference",
                      "parts": [
                        {
                          "error": undefined,
                          "ref": {
                            "$type": "Class",
                            "aliases": [],
                            "annotatingMembers": [],
                            "declaredName": "C",
                            "filters": [],
                            "imports": [],
                            "isSufficient": false,
                            "members": [],
                            "namespaceMembers": [],
                            "prefixes": [],
                            "relationshipMembers": [],
                            "typeRelationships": [],
                          },
                        },
                      ],
                    },
                  },
                ],
              },
              "elements": [],
              "prefixes": [],
              "visibility": "protected",
            },
          ],
          "namespaceMembers": [
            {
              "$type": "OwningMembership",
              "annotations": [],
              "chains": [],
              "element": {
                "$type": "Class",
                "aliases": [],
                "annotatingMembers": [],
                "declaredName": "C",
                "filters": [],
                "imports": [],
                "isSufficient": false,
                "members": [],
                "namespaceMembers": [],
                "prefixes": [],
                "relationshipMembers": [],
                "typeRelationships": [],
              },
              "elements": [],
              "prefixes": [],
              "visibility": "public",
            },
            {
              "$type": "OwningMembership",
              "annotations": [],
              "chains": [],
              "element": {
                "$type": "DataType",
                "aliases": [],
                "annotatingMembers": [],
                "declaredName": "D",
                "filters": [],
                "imports": [],
                "isSufficient": false,
                "members": [],
                "namespaceMembers": [],
                "prefixes": [],
                "relationshipMembers": [],
                "typeRelationships": [],
              },
              "elements": [],
              "prefixes": [],
              "visibility": "private",
            },
          ],
          "prefixes": [],
          "relationshipMembers": [],
        },
        "elements": [],
        "prefixes": [],
      },
    ],
    "prefixes": [],
    "relationshipMembers": [],
  },
}
`;

exports[`namespaces are parseable: toParseKerML 1`] = `
{
  "diagnostics": [],
  "lexerErrors": [],
  "parserErrors": [],
  "value": {
    "$type": "Namespace",
    "aliases": [],
    "annotatingMembers": [],
    "filters": [],
    "imports": [],
    "members": [],
    "namespaceMembers": [
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Namespace",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "N1",
          "declaredShortName": "'1.1'",
          "filters": [],
          "imports": [],
          "members": [],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
        },
        "elements": [],
        "prefixes": [],
      },
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Namespace",
          "aliases": [],
          "annotatingMembers": [
            {
              "$type": "OwningMembership",
              "annotations": [],
              "chains": [],
              "element": {
                "$type": "Documentation",
                "about": [],
                "body": "/* This is an example of a namespace body. */",
                "prefixes": [],
              },
              "elements": [],
              "prefixes": [],
            },
          ],
          "declaredName": "N2",
          "declaredShortName": "'1.2'",
          "filters": [],
          "imports": [],
          "members": [
            {
              "$type": "OwningMembership",
              "annotations": [],
              "chains": [],
              "element": {
                "$type": "Feature",
                "aliases": [],
                "annotatingMembers": [],
                "declaredName": "f",
                "filters": [],
                "imports": [],
                "isNonunique": false,
                "isOrdered": false,
                "isSufficient": false,
                "members": [],
                "namespaceMembers": [],
                "prefixes": [],
                "relationshipMembers": [],
                "typeRelationships": [
                  {
                    "$type": "FeatureTyping",
                    "annotations": [],
                    "chains": [],
                    "elements": [],
                    "prefixes": [],
                    "reference": {
                      "$type": "TypeReference",
                      "parts": [
                        {
                          "error": undefined,
                          "ref": {
                            "$type": "Class",
                            "aliases": [],
                            "annotatingMembers": [],
                            "declaredName": "C",
                            "filters": [],
                            "imports": [],
                            "isSufficient": false,
                            "members": [],
                            "namespaceMembers": [],
                            "prefixes": [],
                            "relationshipMembers": [],
                            "typeRelationships": [],
                          },
                        },
                      ],
                    },
                  },
                ],
              },
              "elements": [],
              "prefixes": [],
            },
          ],
          "namespaceMembers": [
            {
              "$type": "OwningMembership",
              "annotations": [],
              "chains": [],
              "element": {
                "$type": "Class",
                "aliases": [],
                "annotatingMembers": [],
                "declaredName": "C",
                "filters": [],
                "imports": [],
                "isSufficient": false,
                "members": [],
                "namespaceMembers": [],
                "prefixes": [],
                "relationshipMembers": [],
                "typeRelationships": [],
              },
              "elements": [],
              "prefixes": [],
            },
            {
              "$type": "OwningMembership",
              "annotations": [],
              "chains": [],
              "element": {
                "$type": "DataType",
                "aliases": [],
                "annotatingMembers": [],
                "declaredName": "D",
                "filters": [],
                "imports": [],
                "isSufficient": false,
                "members": [],
                "namespaceMembers": [],
                "prefixes": [],
                "relationshipMembers": [],
                "typeRelationships": [],
              },
              "elements": [],
              "prefixes": [],
            },
            {
              "$type": "OwningMembership",
              "annotations": [],
              "chains": [],
              "element": {
                "$type": "Namespace",
                "aliases": [],
                "annotatingMembers": [],
                "declaredName": "N3",
                "filters": [],
                "imports": [],
                "members": [],
                "namespaceMembers": [],
                "prefixes": [],
                "relationshipMembers": [],
              },
              "elements": [],
              "prefixes": [],
            },
          ],
          "prefixes": [],
          "relationshipMembers": [],
        },
        "elements": [],
        "prefixes": [],
      },
    ],
    "prefixes": [],
    "relationshipMembers": [],
  },
}
`;

exports[`namespaces can have comments: toParseKerML 1`] = `
{
  "diagnostics": [],
  "lexerErrors": [],
  "parserErrors": [],
  "value": {
    "$type": "Namespace",
    "aliases": [],
    "annotatingMembers": [],
    "filters": [],
    "imports": [],
    "members": [],
    "namespaceMembers": [
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Namespace",
          "aliases": [],
          "annotatingMembers": [
            {
              "$type": "OwningMembership",
              "annotations": [],
              "chains": [],
              "element": {
                "$type": "Comment",
                "about": [
                  {
                    "$type": "Annotation",
                    "annotatedElement": {
                      "$type": "ElementReference",
                      "parts": [
                        {
                          "error": undefined,
                          "ref": {
                            "$type": "Class",
                            "aliases": [],
                            "annotatingMembers": [],
                            "declaredName": "A",
                            "filters": [],
                            "imports": [],
                            "isSufficient": false,
                            "members": [],
                            "namespaceMembers": [],
                            "prefixes": [],
                            "relationshipMembers": [],
                            "typeRelationships": [],
                          },
                        },
                      ],
                    },
                    "annotations": [],
                    "chains": [],
                    "elements": [],
                    "prefixes": [],
                  },
                ],
                "body": "/* comment about A */",
                "declaredName": "Comment1",
                "prefixes": [],
              },
              "elements": [],
              "prefixes": [],
            },
            {
              "$type": "OwningMembership",
              "annotations": [],
              "chains": [],
              "element": {
                "$type": "Comment",
                "about": [],
                "body": "/* comment about N9 */",
                "declaredName": "Comment2",
                "prefixes": [],
              },
              "elements": [],
              "prefixes": [],
            },
            {
              "$type": "OwningMembership",
              "annotations": [],
              "chains": [],
              "element": {
                "$type": "Comment",
                "about": [],
                "body": "/* also comment about N9 */",
                "prefixes": [],
              },
              "elements": [],
              "prefixes": [],
            },
            {
              "$type": "OwningMembership",
              "annotations": [],
              "chains": [],
              "element": {
                "$type": "Documentation",
                "about": [],
                "body": "/* doc about N9 */",
                "declaredName": "N9_Doc",
                "prefixes": [],
              },
              "elements": [],
              "prefixes": [],
            },
          ],
          "declaredName": "N9",
          "filters": [],
          "imports": [],
          "members": [],
          "namespaceMembers": [
            {
              "$type": "OwningMembership",
              "annotations": [],
              "chains": [],
              "element": {
                "$type": "Class",
                "aliases": [],
                "annotatingMembers": [],
                "declaredName": "A",
                "filters": [],
                "imports": [],
                "isSufficient": false,
                "members": [],
                "namespaceMembers": [],
                "prefixes": [],
                "relationshipMembers": [],
                "typeRelationships": [],
              },
              "elements": [],
              "prefixes": [],
            },
          ],
          "prefixes": [],
          "relationshipMembers": [],
        },
        "elements": [],
        "prefixes": [],
      },
    ],
    "prefixes": [],
    "relationshipMembers": [],
  },
}
`;
